src/utils/price.ts

```diff
@@ -0,0 +1,77 @@
+import BN from 'bn.js'
+import { assert } from './miscellaneous'
+import { TEN, UNLIMITED_ORDER_AMOUNT } from 'const'
+import BigNumber from 'bignumber.js'
+
+export function adjustAmountToLowerPrecision(amount: BN, higherPrecision: number, lowerPrecision: number): BN {
+  if (higherPrecision === lowerPrecision) {
+    // no adjustment required
+    return amount
+  }
+
+  assert(higherPrecision > lowerPrecision, 'higherPrecision must be > lowerPrecision')
+
+  const difference = new BN(higherPrecision - lowerPrecision)
+
+  // divides amount by difference in precision, rounding. E.g:
+  // 1.2345, precision 4, amount == 12345
+  // 1.2345, precision 2, amount == 12345 / 10 ^ (4 - 2) => 12345 / 100 => 123
+  // 1.5550, precision 1, amount == 15550 / 10 ^ (4 - 1) => 15550 / 1000 => 16
+  return amount.divRound(TEN.pow(difference))
+}
+
+function bigNumberToBN(n: BigNumber | BN): BN {
+  if (n instanceof BN) {
+    return n
+  }
+  return new BN(n.integerValue().toString(10))
+}
+
+interface Amounts {
+  buyAmount: BN
+  sellAmount: BN
+}
+
+const MAX = new BigNumber(UNLIMITED_ORDER_AMOUNT.toString())
+const ONE = new BigNumber(1)
+
+/**
+ * Calculates the max amounts within given `spread` for unlimited orders
+ *
+ * Uses BigNumber internally to keep track of decimals
+ * Returns BN for compatibility with ExchangeApi
+ *
+ * @param spread Value between 0 and 100, not inclusive
+ * @param buyTokenPrecision Decimals of buy token
+ * @param sellTokenPrecision Decimals of sell token
+ */
+export function maxAmountsForSpread(spread: number, buyTokenPrecision: number, sellTokenPrecision: number): Amounts {
```

Comment:
nice work with this fn, easy to get tripped up here with the switch between `buyAmount` and `sellAmount` and the tricky BN syntax